src/proto/h1/decode.rs

```diff
@@ -442,19 +478,46 @@ impl ChunkedState {
     fn read_trailer<R: MemRead>(
         cx: &mut Context<'_>,
         rdr: &mut R,
+        trailers_buf: &mut Option<BytesMut>,
     ) -> Poll<Result<ChunkedState, io::Error>> {
         trace!("read_trailer");
-        match byte!(rdr, cx) {
+        let byte = byte!(rdr, cx);
+
+        trailers_buf
+            .as_mut()
+            .expect("trailers_buf is None")
+            .put_u8(byte);
+
+        match byte {
             b'\r' => Poll::Ready(Ok(ChunkedState::TrailerLf)),
             _ => Poll::Ready(Ok(ChunkedState::Trailer)),
         }
     }
+
     fn read_trailer_lf<R: MemRead>(
         cx: &mut Context<'_>,
         rdr: &mut R,
+        trailers_buf: &mut Option<BytesMut>,
+        trailers_cnt: &mut u64,
     ) -> Poll<Result<ChunkedState, io::Error>> {
-        match byte!(rdr, cx) {
-            b'\n' => Poll::Ready(Ok(ChunkedState::EndCr)),
+        let byte = byte!(rdr, cx);
+        match byte {
+            b'\n' => {
+                if *trailers_cnt == usize::MAX as u64 {
+                    return Poll::Ready(Err(io::Error::new(
```

Comment:
Can we write some tests covering this as well?

Reply:
added in [545f8c4](https://github.com/hyperium/hyper/pull/3637/commits/545f8c48d3da589cf7a4154aa6433db1bbe07076)

i also lowered the trailer limit to 1024 as i cannot imagine a use case for sending a million trailers.

Reply:
I added a size limit in [f194be1](https://github.com/hyperium/hyper/pull/3637/commits/f194be13cd6fd15823101a8c775252269136c07c)